src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/evolutionaryAlgorithm/EvolutionaryAlgorithmQueryOptimizer.java

```diff
@@ -0,0 +1,143 @@
+    protected final CostModel costModel;
+    private final int nmCandidates;
+    private final int nmSurvivors;
+    private final int nmSteps;
+    protected final RuleApplicationsOfPlan ruleApplicationCache = new RuleApplicationsOfPlan( new RuleInstances() );
+
+    public EvolutionaryAlgorithmQueryOptimizer( final QueryOptimizationContext ctxt, CostModel costModel,
+                                                final int nmCandidates, final int nmSurvivors, int nmSteps ) {
+        assert ctxt != null;
+
+        this.ctxt = ctxt;
+        this.costModel = costModel;
+        this.nmCandidates = nmCandidates;
+        this.nmSurvivors = nmSurvivors;
+        this.nmSteps = nmSteps;
+    }
+
+    @Override
+    public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws CostEstimationException {
+        final PhysicalPlan initialPhysicalPlan = ctxt.getLogicalToPhysicalPlanConverter().convert( initialPlan, false );
+
+        return optimizePlan( initialPhysicalPlan );
+    }
+
+    public PhysicalPlan optimizePlan( final PhysicalPlan plan ) throws CostEstimationException {
+        // initialize the first generation
+        List<PhysicalPlanWithCost> currentGen = generateTheFirstGen( plan );
+
+        // termination condition of optimizing the plan: number of generations
+        for ( int gen = 1; gen < nmSteps; gen++ ){
+            currentGen = generateTheNextGen( currentGen );
+        }
+        return findPlanWithSmallestCost(currentGen).getPlan();
+    }
+
+    protected List<PhysicalPlanWithCost> generateTheFirstGen( final PhysicalPlan plan ) throws CostEstimationException {
+        List<PhysicalPlan> currentGen = new ArrayList<>();
+        currentGen.add(plan);
+
+        // determine all rule applications for the initial plan
+        Set<RuleApplication> ruleApps = ruleApplicationCache.determineRuleApplications( plan );
+
+        // generate candidates for the first generation
+        final RandomizedSelection<RuleApplication> ruleRandomizedSelect = new RandomizedSelection<>();
+        while ( currentGen.size() < nmCandidates && ruleApps.size() > 0 ){
+            // Pick one rule application based on priority, and apply the selected rule application to the initial plan
+            final RuleApplication ruleApplication = ruleRandomizedSelect.pickOne( ruleApps );
+            currentGen.add( ruleApplication.getResultingPlan() );
+
+            // remove the applied rule application from the set of rule applications
+            ruleApps.remove( ruleApplication );
+        }
+        ruleApplicationCache.replaceRuleApplications( plan, ruleApps);
+
+        return annotatePlansWithCost(currentGen);
+    }
+
+    protected List<PhysicalPlanWithCost> generateTheNextGen( final List<PhysicalPlanWithCost> currentGen ) throws CostEstimationException {
+        List<PhysicalPlan> newCandidates = new ArrayList<>();
+
+        while ( newCandidates.size() < nmCandidates ) {
```

Comment:
Attention: There may not be enough new candidates possible. You need an additional termination criterion for this loop here.